File: android/src/main/java/com/feser/ycapp_foundation/model/notification/TwitchNotification.java
package com.feser.ycapp_foundation.model.notification;

import android.annotation.SuppressLint;
import android.content.Intent;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.TimeZone;

public class TwitchNotification {
	private String id;
	private String channelId;
	private String channelName;
	private String gameId;
	private String gameName;
	private String gameBoxArtUrl;
	private String streamTitle;
	private String summary;
	private Date date;
	private Date startedAt;
	private String creatorNames;
	private List<String> creatorKeys;

	public TwitchNotification(Map<String, String> map) {
		this.id = map.get("id");
		this.channelId = map.get("channelId");
		this.channelName = map.get("channelName");
		this.gameId = map.get("gameId");
		this.gameName = map.get("gameName");
		this.gameBoxArtUrl = map.get("gameBoxArtUrl");
		this.streamTitle = map.get("streamTitle");
		this.date = new Date();
		date.setTime(Long.parseLong(map.get("date")));
		this.startedAt = new Date();
		if (map.containsKey("publishedMills")) {
			startedAt.setTime(Long.parseLong(map.get("publishedMills")));
		}
		if (map.containsKey("summary")) {
			this.summary = map.get("summary");
		}

		if (map.containsKey("creatorName")) {
			this.creatorNames = map.get("creatorNames");
		}
		if (map.containsKey("creatorKeys")) {
			String keyString = map.get("creatorKeys");
			if (keyString != null) {
				String[] keyArray = keyString.split(",");
				this.creatorKeys = new ArrayList<>(Arrays.asList(keyArray));
			}
		}
	}


	public TwitchNotification(Intent intent) {
		this.id = intent.getStringExtra("id");
		this.channelId = intent.getStringExtra("channelId");
		this.channelName = intent.getStringExtra("channelName");
		this.gameId = intent.getStringExtra("gameId");
		this.gameName = intent.getStringExtra("gameName");
		this.gameBoxArtUrl = intent.getStringExtra("gameBoxArtUrl");
		this.streamTitle = intent.getStringExtra("streamTitle");
		this.date = new Date();
		date.setTime(Long.parseLong(intent.getStringExtra("date")));
		this.startedAt = new Date();
		if (intent.hasExtra("publishedMills")) {
			startedAt.setTime(Long.parseLong(intent.getStringExtra("publishedMills")));
		}
		if (intent.hasExtra("summary")) {
			this.summary = intent.getStringExtra("summary");
		}

		if (intent.hasExtra("creatorName")) {
			this.creatorNames = intent.getStringExtra("creatorNames");
		}
		if (intent.hasExtra("creatorKeys")) {
			String keyString = intent.getStringExtra("creatorKeys");
			if (keyString != null) {
				String[] keyArray = keyString.split(",");
				this.creatorKeys = new ArrayList<>(Arrays.asList(keyArray));
			}
		}
	}

	public String getId() {
		return id;
	}

	public String getChannelId() {
		return channelId;
	}

	public String getChannelName() {
		return channelName;
	}

	public String getGameId() {
		return gameId;
	}

	public String getGameName() {
		return gameName;
	}

	public String getGameBoxArtUrl() {
		return gameBoxArtUrl;
	}

	public String getStreamTitle() {
		return streamTitle;
	}

	public String getSummary() {
		if (summary == null) {
			return getChannelName() + " went live!";
		}
		return summary;
	}

	public Date getDate() {
		return date;
	}

	public Date getStartedAt() {
		return startedAt;
	}

	public String formateDate() {
		@SuppressLint("SimpleDateFormat")
		SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
		sdf.setTimeZone(TimeZone.getTimeZone("UTC"));
		return sdf.format(getStartedAt());
	}

	public String formateNow() {
		@SuppressLint("SimpleDateFormat")
		SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
		sdf.setTimeZone(TimeZone.getTimeZone("UTC"));
		return sdf.format(new Date());
	}

	public String getCreatorNames() {
		return creatorNames;
	}

	public List<String> getCreatorKeys() {
		return creatorKeys;
	}

	public int getNotificationId() {
		if (getStartedAt() != null) {
			return (int) (((getDate().getTime() / 1000) + (getStartedAt().getTime() / 1000)) / 2);
		} else {
			return (int) (((getDate().getTime() / 1000)));
		}
	}

	public String getChannelNameLog() {
		return channelName.substring(0, Math.min(30, channelName.length()));
	}

	public String getStreamTitleLog() {
		return streamTitle.substring(0, Math.min(30, streamTitle.length()));
	}

	public boolean isCollab() {
		if (getCreatorNames() == null) {
			return false;
		}
		if (getChannelName().isEmpty()) {
			return false;
		}
		if (getCreatorKeys() == null) {
			return false;
		}
		if (getCreatorKeys().isEmpty()) {
			return false;
		}
		return true;
	}
}

File: lib/model/notification/twitch_notification.dart
class TwitchNotification {
  String id;
  String channelId;
  String channelName;
  String gameId;
  String gameName;
  String gameBoxArtUrl;
  String streamTitle;
  String _summary;
  DateTime date;
  DateTime startedAt;
  String creatorNames;
  List<String> creatorKeys;

  TwitchNotification(
    this.id,
    this.channelId,
    this.channelName,
    this.gameId,
    this.gameName,
    this.gameBoxArtUrl,
    this.streamTitle,
    this._summary,
    this.date,
    this.startedAt,
    this.creatorNames,
    this.creatorKeys,
  );

  factory TwitchNotification.fromMap(Map<String, dynamic> map) {
    String id = map["id"] as String;
    String channelId = map["channelId"] as String;
    String channelName = map["channelName"] as String;
    String gameId = map["gameId"] as String;
    String gameName = map["gameName"] as String;
    String gameBoxArtUrl = map["gameBoxArtUrl"] as String;
    String streamTitle = map["streamTitle"] as String;
    DateTime date =
        DateTime.fromMillisecondsSinceEpoch(int.parse(map["date"] as String));
    DateTime startedAt;
    if (map.containsKey("publishedMills")) {
      startedAt = DateTime.fromMillisecondsSinceEpoch(
          int.parse(map["publishedMills"] as String));
    }
    String summary = map["summary"] as String ?? '';
    String creatorNames = map["creatorNames"] as String ?? '';
    List<String> creatorKeys = (map["creatorKeys"] as String ?? '').split(",");
    return TwitchNotification(
      id,
      channelId,
      channelName,
      gameId,
      gameName,
      gameBoxArtUrl,
      streamTitle,
      summary,
      date,
      startedAt,
      creatorNames,
      creatorKeys,
    );
  }

  String get summery {
    return _summary ?? channelName + " went live!";
  }

  int get notificationId {
    if (startedAt != null) {
      return (((date.millisecondsSinceEpoch ~/ 1000) +
              (startedAt.millisecondsSinceEpoch ~/ 1000)) ~/
          2);
    } else {
      return date.millisecondsSinceEpoch ~/ 1000;
    }
  }

  bool get isCollab {
    if (channelName == null) {
      return false;
    }
    if (channelName.isEmpty) {
      return false;
    }
    if (creatorKeys == null) {
      return false;
    }
    if (creatorKeys.isEmpty) {
      return false;
    }
    return true;
  }
}
